fp-bindgen/src/generators/ts_runtime/mod.rs

```diff
@@ -905,9 +905,10 @@ fn format_struct_fields(fields: &[Field], types: &TypeMap, casing: Casing) -> Ve
     fields
         .iter()
         .flat_map(|field| {
+            let field_optional = field.attrs.skip_serializing_if.is_some();
             let field_decl = match types.get(&field.ty) {
                 Some(Type::Container(name, _)) => {
-                    let optional = if name == "Option" { "?" } else { "" };
+                    let field_optional = field_optional || name == "Option";
```

Comment:
This is a bit tricky now. We could still auto opt-in to skipping the serialization for `Option`s, but then only if the `#[fp(rust_module = "...")]` annotation is *not* used. But that would become rather hard to explain, so I would suggest we just completely drop doing it automatically and leave the decision entirely to the user. Just as Serde does, basically.

It does warrant some explanation of the implications in the CHANGELOG though, which is probably a good moment to pick up FP-2725 right away and use the https://keepachangelog.com/en/1.0.0/ convention here too.

Reply:
Isn't this behaviour what we want though? If it _could be potentially optional_ we just mark it as an optional field. This might be a bit too conservative in some cases but it's not that complicated to reason about, I think?

Reply:
Right, the difference is in whether the optional field should be serialized or not. If a `None` field is serialized it will be explicitly serialized as a `null` value and end up as such on the TypeScript side. If it's not serialized at all, it doesn't go over the wire and thus remains `undefined` on the TypeScript side. So the difference is between whether we want to generate `fieldName: T | null` or `fieldName?: T` in TypeScript.

Thus far, we have always assumed that when you specify `Option<T>` in your protocol, you would be happy to omit the `None` during serialization, and so we would generate `fieldName?: T` for TypeScript. We even enforced this when generating Rust models, by adding the `skip_serializing_if = "Option::is_none"` Serde annotation ourselves.

The problem, where this started to break down, is when Rust models were combined with the `rust_module` annotation. In that case we don't generate the Rust models, and thus we cannot enforce anything. We assumed the user would add the Serde annotation for consistency, but it's just an assumption... they might forget, or they might not even always want that.

So instead of assuming that the presence of an `Option` implies the usage of the `skip_serializing_if` annotation, we should just decouple the two: `Option` without the annotation should generate `fieldName: T | null` as type, while `Option` with the annotation should generate `fieldName?: T` (and also not `fieldName?: T | null`!).

Reply:
@arendjr Do the latest commit (add1fa4957ccf6fd2acc8d7603e503a9c619a7d6) make sense? The non-deno tests are failing since I didn't fix them yet, but do the generated types look like you'd expect?

For reference, the generated `StructWithOptions` now looks like this:
```
export type StructWithOptions = {
    filledString: string | null;
    emptyString: string | null;
    filledRegularOptionString: string | null;
    emptyRegularOptionString: string | null;
    undefinedRegularOptionString?: string;
};
```

Based off of:
```
#[derive(Clone, Debug, Default, Deserialize, PartialEq, Eq, Serialize, Serializable)]
#[fp(rename_all = "camelCase")]
pub struct StructWithOptions {
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub filled_string: String,
    #[serde(default, skip_serializing_if = "String::is_empty")]
    pub empty_string: String,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub filled_regular_option_string: Option<String>,
    #[serde(default, skip_serializing_if = "Option::is_none")]
    pub empty_regular_option_string: Option<String>,
    #[serde(default)]
    pub undefined_regular_option_string: Option<String>,
}
```

Reply:
The TS types from your last comment are exactly opposite what I would expect :sweat_smile:

Reply:
🤦 seems you're right